### ISSUE TITLE
Hang during debugging after restarting debugging while on breakpoint

### ISSUE THREAD

**Issue by mousetraps** [Microsoft/nodejstools#1045]
##### Steps to Reproduce
1. New Node.js Starter Express App
2. Set breakpoint in index route
3. Launch debugging and hit breakpoint in index route
4. Set breakpoint on first LOC in app.js
5. Press "restart debugging" button

VS Hangs :cry: 

- NTVS Version: lastest build from master
- Visual Studio Version: VS2015 U2
- Node.js Version: v6.2.1 x64

**Comment by mjbvz** [comment]
Here's the callstack of the hang I'm seeing:

```
 	[External Code]	
>	Microsoft.NodejsTools.dll!Microsoft.NodejsTools.Project.NodejsProjectLauncher.LaunchDebugger(System.IServiceProvider provider, Microsoft.VisualStudio.Shell.Interop.VsDebugTargetInfo dbgInfo) Line 200	C#
 	Microsoft.NodejsTools.dll!Microsoft.NodejsTools.Project.NodejsProjectLauncher.StartWithDebugger(string startupFile) Line 189	C#
 	Microsoft.NodejsTools.dll!Microsoft.NodejsTools.Project.NodejsProjectLauncher.Start(string file, bool debug) Line 79	C#
 	Microsoft.NodejsTools.dll!Microsoft.NodejsTools.Project.NodejsProjectLauncher.LaunchProject(bool debug) Line 61	C#
 	Microsoft.NodejsTools.dll!Microsoft.VisualStudioTools.Project.CommonProjectConfig.DebugLaunch(uint flags) Line 40	C#
```

I'll see if at also repos with versions of Node < 6.0, since I had to apply #1054 to reproduce this with the listed setup.

**Comment by mjbvz** [comment]
Repos in node versions 5.x and 0.12 as well.

It seems that in this case, the first debugger session does not properly terminate before we start launching the next session, which then hangs VS.

**Comment by mjbvz** [comment]
Continuing the investigation. Here's a quick update:

----

I added a ten second sleep before `LaunchDebugger`, but the hang still happened.

---

My initial thought was that  `node.exe` or the `PressAnyKey.exe` were not being terminated fully between debugging sessions. I also saw some calls to the async request `process.Kill()` without a matching `WaitForExit()`, so I tried adding a bunch of waits. This did nothing.

Then I saw that actually no, `node.exe` and `PressAnyKey.exe` are both killed after the restart, and that the second call to `LaunchDebugger`  relaunches both of them before the hang happens.  

---- 

Another area I investigated was synchronous vs asynchronous debug engine events, thinking that perhaps we sent a kill signal to the debug engine but didn't wait on the response. This also seems to have been a dead end.

----

A few things stood out from this investigation:

* The debugged node programs are being relaunched on the restart.
* This could be a timing issue, but it does not appear to be a race condition regarding when we call `LaunchDebugger`.

Again, it appears to me that the old debug instance is not being cleaned up properly somehow.

As a next step, I am going to investigate start -> Stop flows vs. the restart flows, to see if I can understand where these differ.
